### DISCUSSION TITLE
What does mlflow guarantee in terms of backwards compatibility?

### DISCUSSION THREAD

**Discussion by svbernem** [mlflow/mlflow#9821] [Q&A]
Hi all,

I'm evaluating using mlflow for machine learning lifecycle things, and want to know what kind of backwards compatibility guarantees there are. The only thing I can find regarding this is [this bit in the contributing guidelines](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.md#make-changes-backwards-compatible), but "please think about it" is very different from "you can't commit this", so I wanted to clarify here. My questions are:

- Having an existing mlflow DB, is it guaranteed that this DB can always be used with a newer version of mlflow, preserving all the models, experiments and their metadata? Are there breaking points (major release, minor release)? Are there steps needed to upgrade the DB? Is there a problem if the version-gap becomes too large?
- If an existing mlflow DB is used with a newer version (doing any migration steps that may be necessary), is it guaranteed that the DB still works with an older mlflow version? How old can the version be?
- What backwards compatibility guarantees are made concerning the python API?
- How strictly are these guarantees enforced? Have there been cases in the past where they have been broken, that I could investigate to form a picture for myself?
- Where are these guarantees written down? As I said I only found the paragraph in contribution guidelines. Is there a document that makes these things explicit somewhere?

**Comment by nichoio** [comment]
+1 for this. After checking documentation and Github issues + discussion, it is not clear to me if this project properly follows Semantic Versioning. This makes it difficult for MLflow admins to upgrade between 2.x.x versions without having to worry about whether some API could break.

**Comment by Tsagania** [comment]
What did you do eventually? I need to upgrade from 2.18.0 to 3.5.0 and I think i will download artifacts/params/metrics and relog them creating new experiments/runs

**Comment by B-Step62** [comment]
Hi, we have some high-level policies written in this page: https://mlflow.org/docs/latest/self-hosting/migration/

Please let us know if there are any particular questions should be answered in the documentation. We would like to make server management as easy as possible.
